src/goto-instrument/contracts/instrument_spec_assigns.h

```diff
@@ -0,0 +1,354 @@
+
+  /// Returns the target validity expression for a car_exprt
+  exprt
+  target_validity_expr(const car_exprt &car, bool allow_null_target) const;
+
+  /// Generates the target validity assertion for the given `car` and adds
+  /// it to `dest`.
+  /// The assertion checks that if the car's condition holds then
+  /// the target is `r_ok` (or `NULL` if `allow_null_targets` is true).
+  void target_validity_assertion(
+    const car_exprt &car,
+    bool allow_null_target,
+    goto_programt &dest) const;
+
+  /// Returns inclusion check expression for a single candidate location
+  exprt inclusion_check_single(
+    const car_exprt &lhs,
+    const car_exprt &candidate_car) const;
+
+  /// Returns an inclusion check expression of lhs over all tracked cars.
+  /// \param lhs the lhs expression to check for inclusion
+  /// \param allow_null_lhs if true, allow the lhs to be NULL
+  /// \param include_stack_allocated if true, include stack allocated targets
+  /// in the inclusion check.
+  /// \param cfg_info_opt allows target set pruning if available
+  /// \remark If available, `cfg_info_opt` must point to the `lhs` in question.
+  exprt inclusion_check_full(
+    const car_exprt &lhs,
+    bool allow_null_lhs,
+    bool include_stack_allocated,
+    optionalt<cfg_infot> &cfg_info_opt) const;
+
+  /// Returns an inclusion check assertion of lhs over all tracked cars.
+  /// \param lhs the lhs expression to check for inclusion
+  /// \param allow_null_lhs if true, allow the lhs to be NULL
+  /// \param include_stack_allocated if true, include stack allocated targets
+  /// in the inclusion check.
+  /// \param cfg_info_opt allows target set pruning if available
+  /// \param dest destination program to append instructions to
+  /// \remark If available, `cfg_info_opt` must point to the `lhs` in question.
+  void inclusion_check_assertion(
+    const car_exprt &lhs,
+    bool allow_null_lhs,
+    bool include_stack_allocated,
+    optionalt<cfg_infot> &cfg_info_opt,
+    goto_programt &dest) const;
+
+  /// Generates instructions to invalidate all targets aliased with a
+  /// car that was passed to `free`, assuming the inclusion check passes,
+  /// ie. that the freed_car is fully included in one of the tracked targets.
+  void invalidate_heap_and_spec_aliases(
+    const car_exprt &freed_car,
+    goto_programt &dest) const;
+
+  using conditional_target_sett =
+    std::unordered_set<conditional_target_exprt, irep_hash>;
+
+  /// Set of tracked conditional targets that come from assigns clauses specs
+  conditional_target_sett from_spec_assigns;
+
+  /// Set of tracked conditional targets that come from stack-allocated objects
+  conditional_target_sett from_stack_alloc;
+
+  /// Set of tracked conditional targets that come from heap-allocated objects
+  conditional_target_sett from_heap_alloc;
+
+  using car_mapt = std::
+    unordered_map<const conditional_target_exprt, const car_exprt, irep_hash>;
```

Comment:
I don't think the extra `const` on the map's key is necessary. (But then I've commented elsewhere that I'm not sure whether a `conditional_target_exprt` should be used as a key at all.)